File: app/src/main/java/com/stupidtree/hita/ui/timetable/views/TimeTableViewGroup.java
//package com.stupidtree.hita.ui.timetable.views;
//
//import android.annotation.SuppressLint;
//import android.content.Context;
//import android.content.res.TypedArray;
//import android.graphics.Canvas;
//import android.graphics.Color;
//import android.graphics.DashPathEffect;
//import android.graphics.Paint;
//import android.graphics.Path;
//import android.util.AttributeSet;
//import android.util.TypedValue;
//import android.view.MotionEvent;
//import android.view.View;
//import android.view.ViewGroup;
//import android.view.animation.AnimationUtils;
//import android.view.animation.LayoutAnimationController;
//import java.util.List;
//
//import static com.stupidtree.hita.HITAApplication.HContext;
//import static com.stupidtree.hita.timetable.TimetableCore.DDL;
//
//public class TimeTableViewGroup extends ViewGroup {
//    int week;
//    int width, height;
//    int sectionWidth, sectionHeight = 180;
//    int labelWidth;
//    int labelSize;
//    int labelColor;
//    int timelineColor;
//    boolean firstLoad = true;
//
//    HTime startTime = new HTime(0, 0);
//    BaseActivity activityContext;
//    TimeTableBlockAddView addButton = null;
//    TimeTableBlockView.TimeTablePreferenceRoot root;
//
//    private OnCardClickListener onCardClickListener;
//    private OnCardLongClickListener onCardLongClickListener;
//
//
//    public TimeTableViewGroup(Context context, AttributeSet attrs) {
//        super(context, attrs);
//        typedTimeTableView(attrs, 0);
//    }
//
//    public TimeTableViewGroup(Context context, AttributeSet attrs, int defStyleAttr) {
//        super(context, attrs, defStyleAttr);
//        typedTimeTableView(attrs, defStyleAttr);
//    }
//
//    public void setOnCardClickListener(OnCardClickListener onCardClickListener) {
//        this.onCardClickListener = onCardClickListener;
//    }
//
//    public void setOnCardLongClickListener(OnCardLongClickListener onCardLongClickListener) {
//        this.onCardLongClickListener = onCardLongClickListener;
//    }
//
//    public TimeTableViewGroup(Context context) {
//        super(context);
//    }
//
//    @Override
//    protected void dispatchDraw(Canvas canvas) {
//        try {
//            if (week == TimetableCore.getInstance(HContext).getThisWeekOfTerm())
//                drawTodayRect(canvas);
//            drawLabels(canvas);
//            super.dispatchDraw(canvas);
//        } catch (Exception e) {
//            e.printStackTrace();
//        }
//    }
//
//    private void typedTimeTableView(AttributeSet attrs, int defStyleAttr) {
//        TypedArray a = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.TimeTableViewGroup, defStyleAttr, 0);
//        int n = a.getIndexCount();
//        for (int i = 0; i <= n; i++) {
//            int attr = a.getIndex(i);
//            switch (attr) {
//                //注意获取属性的方式为 styleable的名称_属性名称
//                case R.styleable.TimeTableViewGroup_timeLabelSize:
//                    labelSize = a.getDimensionPixelSize(attr, (int) TypedValue.applyDimension(
//                            TypedValue.COMPLEX_UNIT_SP, 8, getResources().getDisplayMetrics()));
//                    break;
//                case R.styleable.TimeTableViewGroup_timeLabelColor:
//                    labelColor = a.getColor(attr, Color.BLACK);
//                    break;
//                case R.styleable.TimeTableViewGroup_timeLineColor:
//                    // 默认颜色设置为黑色
//                    timelineColor = a.getColor(attr, Color.BLACK);
//                    break;
//                case R.styleable.TimeTableViewGroup_timeLabelWidth:
//                    labelWidth = a.getDimensionPixelSize(attr, (int) TypedValue.applyDimension(
//                            TypedValue.COMPLEX_UNIT_DIP, 24, getResources().getDisplayMetrics()));
//                    break;
//
//
//            }
//        }
//
//    }
//
//    private void drawTodayRect(Canvas canvas) {
//        int left = labelWidth + sectionWidth * (TimetableCore.getDOW(TimetableCore.getNow()) - 1);
//        int right = (left + sectionWidth);
//        Paint paint = new Paint();
//        paint.setColor(root.getTodayBGColor());
//        canvas.drawRect(left, 0, right, height, paint);
//    }
//
//    private void drawLabels(Canvas canvas) {
//        Paint paint = new Paint();
//        paint.setColor(labelColor);
//        paint.setTextSize(labelSize);
//        paint.setTextAlign(Paint.Align.LEFT);
//
//
//        DashPathEffect pathEffect = new DashPathEffect(new float[]{20f, 20f}, 0f);
//        Paint paint2 = new Paint();
//        paint2.setStyle(Paint.Style.STROKE);
//        paint2.setStrokeWidth(1);
//        paint2.setColor(timelineColor);
//        paint2.setPathEffect(pathEffect);
//
//
//        for (int i = startTime.hour; i <= 23; i++) {
//            HTime temp = new HTime(i, 0);
//            int left = 0;
//            //int right = (left + labelWidth);
//            float x = startTime.before(temp) ? 1.0f : -1.0f;
//            float startTimeFromBeginning = x * startTime.getDuration(temp);
//            int top = (int) ((startTimeFromBeginning / 60f) * sectionHeight);
//            //nt bottom = top + sectionHeight;
//            canvas.drawText(temp.tellTime(), left, top + labelSize, paint);
//            if (root.drawBGLine()) {
//                Path path = new Path();
//                path.moveTo(0, top);
//                path.lineTo(width, top + 1);
//                canvas.drawPath(path, paint2);
//            }
//
//            // label.layout(left, top, right, bottom);
//        }
//
//    }
//
//    public void notifyRefresh() {
//        this.sectionHeight = root.getCardHeight();
//        this.startTime = root.getStartTime();
//        if (root.animEnabled() && firstLoad) {
//            firstLoad = false;
//            setLayoutAnimation(new LayoutAnimationController(AnimationUtils.loadAnimation(
//                    getContext(), R.anim.recycler_animation_float_up), 0.08f));
//        } else {
//            setLayoutAnimation(null);
//        }
//    }
//
//    @SuppressLint("ClickableViewAccessibility")
//    @Override
//    public boolean onTouchEvent(MotionEvent event) {
//        //
//        // Log.e("action", String.valueOf(event.getAction()));
//        if (event.getAction() == MotionEvent.ACTION_UP) {
//            removeView(addButton);
//            int dow = (int) event.getX() / sectionWidth + 1;
//            HTime time = startTime.getAdded((int) (event.getY() / sectionHeight * 60f));
//            TimePeriod period = TimetableCore.getClassSimplifiedTimeByTimeContainedIn(time);
//            if (period != null) {
//                if (period.start.before(startTime)) period.setStart(startTime);
//                addButton = new TimeTableBlockAddView(activityContext, week, dow,
//                        period
//                );
//                addView(addButton);
//
//            }
//        } else {
//            removeView(addButton);
//        }
//        return super.onTouchEvent(event);
//    }
//
//    @Override
//    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
//        //Log.e("width,height",widthMeasureSpec+","+heightMeasureSpec);
//        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
//        @SuppressLint("DrawAllocation") int totalMinutes = root.getStartTime().getDuration(new HTime(23, 59));
//        setMeasuredDimension(MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.makeMeasureSpec((int) ((float) totalMinutes / 60f * sectionHeight), MeasureSpec.EXACTLY));
//        width = MeasureSpec.getSize(widthMeasureSpec);
//        height = MeasureSpec.makeMeasureSpec((int) ((float) totalMinutes / 60f * sectionHeight), MeasureSpec.EXACTLY);
//        sectionWidth = (width - labelWidth) / 7;
//        //setMeasuredDimension(width, height);
//        final int count = getChildCount();
//        for (int i = 0; i < count; i++) {
//            View child = getChildAt(i);
//            if (child instanceof TimeTableBlockView) {
//                int cw = MeasureSpec.makeMeasureSpec(sectionWidth, MeasureSpec.EXACTLY);
//                int cH;
//                EventItem eventItem = ((TimeTableBlockView) child).getEvent();
//                if (eventItem.eventType != DDL) {
//                    cH = MeasureSpec.makeMeasureSpec(getCardHeight((TimeTableBlockView) child), MeasureSpec.EXACTLY);
//                } else {
//                    cH = MeasureSpec.makeMeasureSpec((int) ((18 / 60f) * sectionHeight), MeasureSpec.EXACTLY);
//                }
//                child.measure(cw, cH);
//            } else if (child instanceof TimeTableNowLine) {
//                int cw = MeasureSpec.makeMeasureSpec(sectionWidth, MeasureSpec.EXACTLY);
//                int cH = MeasureSpec.makeMeasureSpec(4, MeasureSpec.EXACTLY);
//                child.measure(cw, cH);
//            } else if (child instanceof TimeTableBlockAddView) {
//                int cw = MeasureSpec.makeMeasureSpec(sectionWidth, MeasureSpec.EXACTLY);
//                int cH = MeasureSpec.makeMeasureSpec((int) ((((TimeTableBlockAddView) child).getDuration() / 60f) * sectionHeight), MeasureSpec.EXACTLY);
//                child.measure(cw, cH);
//            } else {
//                this.measureChild(child, widthMeasureSpec, heightMeasureSpec);
//            }
//
//            //
//        }
//    }
//
//    public void init(BaseActivity activity, int week, TimeTableBlockView.TimeTablePreferenceRoot root) {
//        this.week = week;
//        this.activityContext = activity;
//        this.root = root;
//        this.sectionHeight = root.getCardHeight();
//        this.startTime = root.getStartTime();
//        setClickable(true); //设置为可点击，否则onTouchEvent只返回DOWN
//    }
//
//    @Override
//    protected void onLayout(boolean changed, int l, int t, int r, int b) {
//        int count = getChildCount();//获得子控件个数
//        for (int i = 0; i < count; i++) {
//            View child = getChildAt(i);
//            if (child instanceof TimeTableBlockView) {
//                TimeTableBlockView block = (TimeTableBlockView) child;
//                float lastTime = block.getDuration();
//                float x = startTime.before(block.getEvent().startTime) ? 1.0f : -1.0f;
//                float startTimeFromBeginning = x * startTime.getDuration(block.getEvent().startTime);
//                int courseInWeek = block.getDow() - 1;//获得周几
//                //计算左边的坐标
//                int left = labelWidth + (sectionWidth * courseInWeek);
//                //计算右边坐标
//                int right = (left + sectionWidth);
//                //计算顶部坐标
//                int top = (int) ((startTimeFromBeginning / 60f) * sectionHeight);
//                //计算底部坐标
//                if (((TimeTableBlockView) child).getEvent().eventType == DDL) {
//                    lastTime = 18;
//                }
//                int bottom = top + (int) ((lastTime / 60f) * sectionHeight);
//                // block.measure(sectionWidth,getCardHeight(block));
//                block.layout(left, top, right, bottom);
//            } else if (child instanceof TimeTableNowLine) {
//                @SuppressLint("DrawAllocation") float startTimeFromBeginning = startTime.getDuration(new HTime(TimetableCore.getNow()));
//                int top = (int) ((startTimeFromBeginning / 60f) * sectionHeight);
//                child.layout(0, top, width, top + 4);
//            } else if (child instanceof TimeTableBlockAddView) {
//                TimeTableBlockAddView add = (TimeTableBlockAddView) child;
//                int left = sectionWidth * (add.getDow() - 1) + labelWidth;
//                int right = (left + sectionWidth);
//                float startTimeFromBeginning = startTime.getDuration(add.getTimePeriod().start);
//                int top = (int) ((startTimeFromBeginning / 60f) * sectionHeight);
//                int bottom = top + (int) ((add.getDuration() / 60f) * sectionHeight);
//                //Log.e("pos",top+","+bottom+","+left+","+right);
//                add.layout(left, top, right, bottom);
//            }
//
//
//            // Log.e("mes:",top+","+bottom+","+left+","+right);
//        }
//    }
//
//    public void addBlock(Object o) {
//        if (o instanceof EventItem) {
//
//            TimeTableBlockView timeTableBlockView = new TimeTableBlockView(getContext(), o, root);
//            if (onCardClickListener != null)
//                timeTableBlockView.setOnCardClickListener(new TimeTableBlockView.OnCardClickListener() {
//                @Override
//                public void OnClick(View v, EventItem ei) {
//                    onCardClickListener.onEventClick(v, ei);
//                }
//            });
//            if (onCardLongClickListener != null)
//                timeTableBlockView.setOnCardLongClickListener(new TimeTableBlockView.OnCardLongClickListener() {
//                @Override
//                public boolean OnLongClick(final View v, final EventItem ei) {
//                    return onCardLongClickListener.onEventLongClick(v, ei);
//                }
//            });
//            addView(timeTableBlockView);
//        } else if (o instanceof List) {
//            TimeTableBlockView timeTableBlockView = new TimeTableBlockView(getContext(), o, root);
//            if (onCardClickListener != null)
//                timeTableBlockView.setOnDuplicateCardClickListener(new TimeTableBlockView.OnDuplicateCardClickListener() {
//                @Override
//                public void OnDuplicateClick(View v, final List<EventItem> list) {
//                    onCardClickListener.onDuplicateEventClick(v, list);
//                }
//                });
//            if (onCardLongClickListener != null)
//                timeTableBlockView.setOnDuplicateCardLongClickListener(new TimeTableBlockView.OnDuplicateCardLongClickListener() {
//                    @Override
//                    public boolean OnDuplicateLongClick(View v, List<EventItem> list) {
//                        return onCardLongClickListener.onDuplicateEventClick(v, list);
//                }
//            });
//            addView(timeTableBlockView);
//        }
//
//    }
//
//    public interface OnCardClickListener {
//        void onEventClick(View v, EventItem eventItem);
//
//        void onDuplicateEventClick(View v, List<EventItem> eventItems);
//    }
//
//
//    private int getCardHeight(TimeTableBlockView timeTableBlockView) {
//        return (int) ((timeTableBlockView.getDuration() / 60f) * sectionHeight);
//    }
//
//
//    public interface OnCardLongClickListener {
//        boolean onEventLongClick(View v, EventItem eventItem);
//
//        boolean onDuplicateEventClick(View v, List<EventItem> eventItems);
//    }
//
//
//}

File: app/src/main/java/com/stupidtree/hita/ui/timetable/views/TimeTableViewGroup.kt
package com.stupidtree.hita.ui.timetable.views

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.view.animation.LayoutAnimationController
import com.stupidtree.hita.R
import com.stupidtree.hita.data.model.timetable.EventItem
import com.stupidtree.hita.data.model.timetable.TimeInDay
import com.stupidtree.hita.ui.timetable.views.TimeTableBlockView.*
import com.stupidtree.hita.utils.TimeUtils

class TimeTableViewGroup : ViewGroup {
    var mWidth = 0
    var mHeight = 0
    var sectionWidth = 0
    var sectionHeight = 180
    var labelWidth = 0
    var labelSize = 0
    var labelColor = 0
    var timelineColor = 0
    var firstLoad = true
    var startTime: TimeInDay = TimeInDay(0, 0)
    var endTime = TimeInDay(24, 0)
    var addButton: TimeTableBlockAddView? = null
    var root: TimeTablePreferenceRoot? = null
    private var onCardClickListener: OnCardClickListener? = null
    private var onCardLongClickListener: OnCardLongClickListener? = null
    private val mLabelPaint = Paint()
    private val mPathEffect = DashPathEffect(floatArrayOf(20f, 20f), 0f)
    private val mLinePaint = Paint()
    private val mLinePath = Path()

    constructor(context: Context?, attrs: AttributeSet) : super(context, attrs) {
        typedTimeTableView(attrs, 0)
    }

    constructor(context: Context?, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        typedTimeTableView(attrs, defStyleAttr)
    }

    fun setOnCardClickListener(onCardClickListener: OnCardClickListener?) {
        this.onCardClickListener = onCardClickListener
    }

    fun setOnCardLongClickListener(onCardLongClickListener: OnCardLongClickListener?) {
        this.onCardLongClickListener = onCardLongClickListener
    }

    constructor(context: Context?) : super(context) {}

    override fun dispatchDraw(canvas: Canvas) {
//            if (week == TimetableCore.getInstance(HContext).getThisWeekOfTerm())
//                drawTodayRect(canvas);
        drawLabels(canvas)
        super.dispatchDraw(canvas)
    }

    private fun typedTimeTableView(attrs: AttributeSet, defStyleAttr: Int) {
        val a = context.theme.obtainStyledAttributes(
            attrs,
            R.styleable.TimeTableViewGroup,
            defStyleAttr,
            0
        )
        val n = a.indexCount
        for (i in 0..n) {
            when (val attr = a.getIndex(i)) {
                R.styleable.TimeTableViewGroup_timeLabelSize -> labelSize = a.getDimensionPixelSize(
                    attr, TypedValue.applyDimension(
                        TypedValue.COMPLEX_UNIT_SP, 8f, resources.displayMetrics
                    ).toInt()
                )
                R.styleable.TimeTableViewGroup_timeLabelColor -> labelColor =
                    a.getColor(attr, Color.BLACK)
                R.styleable.TimeTableViewGroup_timeLineColor ->                     // 默认颜色设置为黑色
                    timelineColor = a.getColor(attr, Color.BLACK)
                R.styleable.TimeTableViewGroup_timeLabelWidth -> labelWidth =
                    a.getDimensionPixelSize(
                        attr, TypedValue.applyDimension(
                            TypedValue.COMPLEX_UNIT_DIP, 24f, resources.displayMetrics
                        ).toInt()
                    )
            }
        }
    }

    private fun drawTodayRect(canvas: Canvas) {
        val left: Int = labelWidth + sectionWidth * (TimeUtils.currentDOW() - 1)
        val right = left + sectionWidth
        val paint = Paint()
        paint.color = root!!.todayBGColor
        canvas.drawRect(left.toFloat(), 0f, right.toFloat(), mHeight.toFloat(), paint)
    }


    private fun drawLabels(canvas: Canvas) {
        mLabelPaint.color = labelColor
        mLabelPaint.textSize = labelSize.toFloat()
        mLabelPaint.textAlign = Paint.Align.LEFT
        mLinePaint.style = Paint.Style.STROKE
        mLinePaint.strokeWidth = 1f
        mLinePaint.color = timelineColor
        mLinePaint.pathEffect = mPathEffect
        for (i in startTime.hour..23) {
            val temp = TimeInDay(i, 0)
            val left = 0
            val x = if (startTime.before(temp)) 1.0f else -1.0f
            val startTimeFromBeginning: Float = x * startTime.getDistanceInMinutes(temp)
            val top = (startTimeFromBeginning / 60f * sectionHeight).toInt()
            canvas.drawText(temp.toString(), left.toFloat(), (top + labelSize).toFloat(),mLabelPaint)
            if (root!!.drawBGLine()) {
                mLinePath.moveTo(0f, top.toFloat())
                mLinePath.lineTo(mWidth.toFloat(), (top + 1).toFloat())
                canvas.drawPath(mLinePath, mLinePaint)
            }
        }
    }

    fun notifyRefresh() {
        sectionHeight = root!!.cardHeight
        startTime = root!!.startTime
        if (root!!.animEnabled() && firstLoad) {
            firstLoad = false
            layoutAnimation = LayoutAnimationController(
                AnimationUtils.loadAnimation(
                    context, R.anim.recycler_animation_float_up
                ), 0.08f
            )
        } else {
            layoutAnimation = null
        }
    }

//    @SuppressLint("ClickableViewAccessibility")
//    override fun onTouchEvent(event: MotionEvent): Boolean {
//        if (event.action == MotionEvent.ACTION_UP) {
//            removeView(addButton)
//            val dow = event.x.toInt() / sectionWidth + 1
//            val time: TimeInDay = startTime.getAdded((event.y / sectionHeight * 60f).toInt())
//            val period: TimePeriod = TimetableCore.getClassSimplifiedTimeByTimeContainedIn(time)
//            if (period != null) {
//                if (period.start.before(startTime)) period.setStart(startTime)
//                addButton = TimeTableBlockAddView(activityContext!!, week, dow,
//                        period
//                )
//                addView(addButton)
//            }
//        } else {
//            removeView(addButton)
//        }
//        return super.onTouchEvent(event)
//    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val totalMinutes: Int = root!!.startTime.getDistanceInMinutes(endTime)
        setMeasuredDimension(
            MeasureSpec.getSize(widthMeasureSpec),
            MeasureSpec.makeMeasureSpec(
                (totalMinutes.toFloat() / 60f * sectionHeight).toInt(),
                MeasureSpec.EXACTLY
            )
        )
        mWidth = MeasureSpec.getSize(widthMeasureSpec)
        mHeight = MeasureSpec.makeMeasureSpec(
            (totalMinutes.toFloat() / 60f * sectionHeight).toInt(),
            MeasureSpec.EXACTLY
        )
        sectionWidth = (mWidth - labelWidth) / 7
        //setMeasuredDimension(width, height);
        for (i in 0 until childCount) {
            when (val child = getChildAt(i)) {
                is TimeTableBlockView -> {
                    val cw = MeasureSpec.makeMeasureSpec(sectionWidth, MeasureSpec.EXACTLY)
                    val cH: Int =
                        MeasureSpec.makeMeasureSpec(getCardHeight(child), MeasureSpec.EXACTLY)
                    child.measure(cw, cH)
                }
                is TimeTableNowLine -> {
                    val cw = MeasureSpec.makeMeasureSpec(sectionWidth, MeasureSpec.EXACTLY)
                    val cH = MeasureSpec.makeMeasureSpec(4, MeasureSpec.EXACTLY)
                    child.measure(cw, cH)
                }
                is TimeTableBlockAddView -> {
                    val cw = MeasureSpec.makeMeasureSpec(sectionWidth, MeasureSpec.EXACTLY)
                    val cH = MeasureSpec.makeMeasureSpec(
                        (child.duration / 60f * sectionHeight).toInt(),
                        MeasureSpec.EXACTLY
                    )
                    child.measure(cw, cH)
                }
                else -> {
                    measureChild(child, widthMeasureSpec, heightMeasureSpec)
                }
            }

            //
        }
    }

    fun init(root: TimeTablePreferenceRoot) {
        this.root = root
        sectionHeight = root.cardHeight
        startTime = root.startTime
        isClickable = true //设置为可点击，否则onTouchEvent只返回DOWN
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        val count = childCount //获得子控件个数
        for (i in 0 until count) {
            when (val child = getChildAt(i)) {
                is TimeTableBlockView -> {
                    val lastTime = child.getDuration().toFloat()
                    val startTimeFromBeginning: Int =
                        startTime.getDistanceInMinutes(child.getEvent().from.time)
                    val courseInWeek = child.getDow() - 1 //获得周几
                    //计算左边的坐标
                    val left = labelWidth + sectionWidth * courseInWeek
                    //计算右边坐标
                    val right = left + sectionWidth
                    //计算顶部坐标
                    val top = (startTimeFromBeginning / 60f * sectionHeight).toInt()
                    val bottom = top + (lastTime / 60f * sectionHeight).toInt()
                    // block.measure(sectionWidth,getCardHeight(block));
                    child.layout(left, top, right, bottom)
                }
                is TimeTableNowLine -> {
                    val startTimeFromBeginning =
                        startTime.getDistanceInMinutes(System.currentTimeMillis())
                    val top = (startTimeFromBeginning / 60f * sectionHeight).toInt()
                    child.layout(0, top, mWidth, top + 4)
                }
//                is TimeTableBlockAddView -> {
//                    val left: Int = sectionWidth * (child.getDow() - 1) + labelWidth
//                    val right = left + sectionWidth
//                    val startTimeFromBeginning: Float = startTime.getDuration(child.timePeriod.start)
//                    val top = (startTimeFromBeginning / 60f * sectionHeight).toInt()
//                    val bottom = top + (child.duration / 60f * sectionHeight).toInt()
//                    //Log.e("pos",top+","+bottom+","+left+","+right);
//                    child.layout(left, top, right, bottom)
//                }
            }


            // Log.e("mes:",top+","+bottom+","+left+","+right);
        }
    }

    fun addBlock(o: Any) {
        if (o is EventItem) {
            val timeTableBlockView = TimeTableBlockView(context, o, root!!)
            if (onCardClickListener != null) timeTableBlockView.onCardClickListener =
                object : TimeTableBlockView.OnCardClickListener {
                    override fun OnClick(v: View, ei: EventItem) {
                        onCardClickListener!!.onEventClick(v, ei)
                    }
                }
            if (onCardLongClickListener != null) timeTableBlockView.onCardLongClickListener =
                object : TimeTableBlockView.OnCardLongClickListener {
                    override fun OnLongClick(v: View, ei: EventItem): Boolean {
                        return onCardLongClickListener!!.onEventLongClick(v, ei)
                    }
                }
            addView(timeTableBlockView)
        } else if (o is List<*>) {
            val timeTableBlockView = TimeTableBlockView(context, o, root!!)
            if (onCardClickListener != null) timeTableBlockView.onDuplicateCardClickListener =
                object : OnDuplicateCardClickListener {
                    override fun OnDuplicateClick(v: View, list: List<EventItem>) {
                        onCardClickListener!!.onDuplicateEventClick(v, list)
                    }
                }
            if (onCardLongClickListener != null) timeTableBlockView.onDuplicateCardLongClickListener =
                object : OnDuplicateCardLongClickListener {
                    override fun onDuplicateLongClick(v: View, list: List<EventItem>): Boolean {
                        return onCardLongClickListener!!.onDuplicateEventClick(v, list)
                    }
                }
            addView(timeTableBlockView)
        }
    }

    interface OnCardClickListener {
        fun onEventClick(v: View, eventItem: EventItem)
        fun onDuplicateEventClick(v: View, eventItems: List<EventItem>)
    }

    private fun getCardHeight(timeTableBlockView: TimeTableBlockView): Int {
        return (timeTableBlockView.getDuration() / 60f * sectionHeight).toInt()
    }

    interface OnCardLongClickListener {
        fun onEventLongClick(v: View, eventItem: EventItem): Boolean
        fun onDuplicateEventClick(v: View, eventItems: List<EventItem>): Boolean
    }
}